subsys/bluetooth/services/cgms/cgms.c

```diff
@@ -0,0 +1,455 @@
+ *
+ * SPDX-License-Identifier: LicenseRef-Nordic-5-Clause
+ */
+
+#include <zephyr/types.h>
+#include <stddef.h>
+#include <string.h>
+#include <errno.h>
+#include <zephyr/zephyr.h>
+#include <zephyr/init.h>
+#include <zephyr/net/buf.h>
+
+#include <zephyr/bluetooth/bluetooth.h>
+#include <zephyr/bluetooth/hci.h>
+#include <zephyr/bluetooth/conn.h>
+#include <zephyr/bluetooth/uuid.h>
+#include <zephyr/bluetooth/gatt.h>
+#include <bluetooth/services/cgms.h>
+#include "cgms_internal.h"
+
+#include <zephyr/logging/log.h>
+
+LOG_MODULE_REGISTER(cgms, CONFIG_BT_CGMS_LOG_LEVEL);
+
+#define SEC_PER_DAY				86400
+
+#define CGMS_SVC_MEAS_ATTR_IDX			1
+#define CGMS_SVC_RACP_ATTR_IDX			12
+#define CGMS_SVC_SOCP_ATTR_IDX			15
+
+#define CGMS_FEAT_LENGTH			6
+#define CGMS_MEAS_LENGTH			15
+#define CGMS_STATUS_LENGTH			5
+#define CGMS_SST_LENGTH				9
+
+static struct k_work_delayable report_meas_work;
+
+/* We instantiate a CGMS server, and use it to hold the parameters that describe its state.
+ * Each device should only have one instance according to the specification.
+ */
+static struct cgms_server cgms_inst;
+
+static void cgms_meas_ccc_cfg_changed(const struct bt_gatt_attr *attr, uint16_t value)
+{
+	LOG_INF("CGMS Measurement: notification %s",
+			value == BT_GATT_CCC_NOTIFY ? "enabled" : "disabled");
+}
+
+static ssize_t read_feature(struct bt_conn *conn, const struct bt_gatt_attr *attr,
+			void *buf, uint16_t len, uint16_t offset)
+{
+	struct net_buf_simple *feature_buf = NET_BUF_SIMPLE(CGMS_FEAT_LENGTH);
+
+	net_buf_simple_init(feature_buf, 0);
+
+	net_buf_simple_add_le24(feature_buf, cgms_inst.feature.feature);
+	net_buf_simple_add_u8(feature_buf,
+		(cgms_inst.feature.type & 0xF) |
+		((cgms_inst.feature.sample_location & 0xF) << 4));
+	net_buf_simple_add_le16(feature_buf, 0xFFFF);
+
+	return bt_gatt_attr_read(conn, attr, buf, len, offset, feature_buf->data, feature_buf->len);
+}
+
+static ssize_t read_session_start_time(struct bt_conn *conn, const struct bt_gatt_attr *attr,
+			void *buf, uint16_t len, uint16_t offset)
+{
+	struct net_buf_simple *sst_buf = NET_BUF_SIMPLE(CGMS_SST_LENGTH);
+
+	net_buf_simple_init(sst_buf, 0);
+
+	net_buf_simple_add_le16(sst_buf, cgms_inst.sst.year);
+	net_buf_simple_add_u8(sst_buf, cgms_inst.sst.month);
+	net_buf_simple_add_u8(sst_buf, cgms_inst.sst.day);
+	net_buf_simple_add_u8(sst_buf, cgms_inst.sst.hour);
+	net_buf_simple_add_u8(sst_buf, cgms_inst.sst.minute);
+	net_buf_simple_add_u8(sst_buf, cgms_inst.sst.second);
+	net_buf_simple_add_u8(sst_buf, cgms_inst.sst.timezone);
+	net_buf_simple_add_u8(sst_buf, cgms_inst.sst.dst_offset);
+
+	return bt_gatt_attr_read(conn, attr, buf, len, offset, sst_buf->data, sst_buf->len);
+}
+
+static bool leap_year_check(uint16_t year)
+{
+	if (year % 400 == 0) {
+		return true;
+	} else if (year % 100 == 0) {
+		return false;
+	} else if (year % 4 == 0) {
+		return true;
+	} else {
+		return false;
+	}
+}
+
+static int cal_sst_offset(struct net_buf_simple *sst_buffer)
```

Comment:
Please test this function in runtime (especially the calculation of the yyyy:mm:dd  part which will execute very rarely).